Added common remediation page about common a11y issues (#335)
springernature/frontend-playbook: accessibility/common-remediations.md

Built-in interactive HTML elements like anchors, text fields, buttons, and select lists (among others) are handled automatically by the browser. Users can interact with them with the mouse, the keyboard, by voice, or any other kind of technology that they like. 

Elements with no semantics (divs and spans) don’t have this automatic handling, and a simple `onclick` handler doesn’t provide it. These elements generally can’t be interacted with by users of Assistive Technology unless you write large amounts of JavaScript to make it possible. 

Where a built-in interactive element is available, use it. 

We do this:
```html
<button>Toggle UI</button>
<a href=”/”>Visit this location</a>
```
We _don’t_ do this:
```html
<div onclick=”doFunc();”>Toggle UI</div>
<span>Visit this location</a>
```

#### Links with no href

All anchor links (the `a` element) require an `href` attribute in order to be functional as links. Omitting the `href` attribute prevents keyboard users and voice input users from being able to use the link at all. 

We do this:
```html
<a href=”#location”>Visit this location</a>
```
We _don’t_ do this:
```html
<a onclick=”myFunc();”>Visit this location</a>
```

> ####  Anchor elements with the [tabindex=-1] attribute are not usable by keyboard users.

Reviewer: ```suggestion
####  Inappropriate negative tabindex

Interactive elements with a negative tabindex attribute (`tabindex=-1`) are not usable by keyboard users. 

```

I've generalised this a little because it could be applied incorrectly to a button as well as a link

> ### Visual order is not followed with source code

Reviewer: This section is about focus order, so I think it'd make sense if it was moved above the keyboard navigation heading, so it comes just after the visible focus styles section 😄

We do this:
```html
<img src="/images/kinghenry.jpg" alt="King Henry VIII of England" />
```

We don't do this:
```html
 <img src="/images/kinghenry.jpg" alt="Image of King" />
```

### Meaningless images with meaningful alt attributes

If an image is decorative, the alt attribute should be present, but its value left blank. This conveys that the image isn’t necessary for understanding the page.

We do this:
```html
 <img src="/images/spacer.gif" alt="" />
 <img src="/images/bg_img.gif" alt="" />
```

We _don’t_ do this:
```html
 <img src="/images/spacer.gif" alt="spacer image" />
 <img src="/images/bg_img.gif" alt="background image" />
```

### Redundant alt attributes
If an image represents information that is already available in text, the alt attribute should be present, but its value left blank. This prevents screen reader users from having to listen to multiple repetitions of the same information. 

We do this:
```html
<div class=”person”><img src="/images/mrs-a-person.gif" alt=""><a href=”/”>Mrs. A. Person</a></div>
```

We _don’t_ do this:
```html
<div class=”person”><img src="/images/mrs-a-person.gif" alt="Mrs. A. Person" /><a href=”/”>Mrs. A. Person</a></div>
```

> ### Logo image

Reviewer: ```suggestion
### Brand logos
```

> For an image `alt` attribute is sufficient and more meaningful.

Reviewer: Let's split all of these items into individual headings, with one concept per heading. 

When we do accessibility reviews, we'll be linking to this document in our reports whenever we find a problem. So we want to have reduced cases on this page that demonstrate one problem at a time, so people know exactly what we want them to fix.

Author: ok

Author: Will resolve this with our coming iterations of this document

No need of title attribute, use an inline image with an alt attribute for the brand logo.
Make sure any text in images of text is at least 14 points and has good contrast with the background.

We do this:
```html

> <a href="https://www.w3.org/"> <img src="w3c.png" alt="W3C home"></a>

Reviewer: ```suggestion
<a href="https://www.w3.org/"> <img src="w3c.png" alt="W3C"></a>
```

Alt text must always match the text in the image exactly - no words should be added before or after.  We need a separate heading for this rule, too!

# Common remediations

This page describes some common accessibility issues, and gives examples and suggestions for repairs. 

> 

Reviewer: Can we have a table of contents too? Will make it easier for us to see if the order of the sections makes sense 😬

Author: yes actually I was going to add it once we have most of the points covered

We do this:
```html
<button>Toggle UI</button>
<a href=”/”>Visit this location</a>
```
We _don’t_ do this:
```html
<div onclick=”doFunc();”>Toggle UI</div>
<span>Visit this location</a>
```

#### Links with no href

All anchor links (the `a` element) require an `href` attribute in order to be functional as links. Omitting the `href` attribute prevents keyboard users and voice input users from being able to use the link at all. 

We do this:
```html
<a href=”#location”>Visit this location</a>
```
We _don’t_ do this:
```html
<a onclick=”myFunc();”>Visit this location</a>
```

####  Inappropriate negative tabindex

Interactive elements with a negative tabindex attribute (`tabindex=-1`) are not usable by keyboard users. 

### Visual order is not followed with source code
Just try to follow the visual order of elements while coding in HTML, don’t adjust elements to take positions with CSS, then this is confusing to keyboard users who expect to tab forwards and backwards in a linear sequence.

As a rule, try tabbing through your pages every so often just to make sure you haven't accidentally messed up the tab order. 

We do this:
```html

> <button> I Should <button />

Reviewer: ```suggestion
<button> I Should </button>
```

he he 😄

<button> Be focused<button />
<button> Last <button />
```

We don't do this:
```html
<button style=”float: right”> I Should <button />
<button> Be focused<button />
<button> Last <button />
```

> ### No advance warning when opening new window 

Reviewer: ```suggestion
## No advance warning when opening new window 
```
Moves this into its own heading group. I might fold it into a new group later that covers context changes in general, if I can think of a way that makes sense 🤔

Opening new windows automatically when a link is activated can be disorienting for people who have difficulty perceiving visual content, and for some people with cognitive disabilities, if they are not warned in advance.

We do this:
```html
<a href="https://mysite.com" target="_blank" rel="noopener">
 My site
 <span class="visuallyhidden">, opens in a new window</span>
</a>
```

We don’t do this:
```html
<a href="knitting.html" target="_blank">All about Knitting</a>
```

## Images

### Images with no alt attribute

Alt attributes are required on all `img` tags. Your HTML is invalid without them. 

We do this:
```html
<img src="/images/NYCpark.png" alt="Aerial view of Central Park in New York" />
```

We don't do this:
```html
<img src="/images/NYCpark.png" />
```

### Meaningful images with meaningless alt attributes

Alt text should be used to convey the intent of an image. Avoid using generic strings like “photo”, “image”, or “icon” as alt values, as they don’t communicate valuable content to the user. Be as descriptive as possible.

We do this:
```html

> <img src="/images/kinghenry.jpg" alt="King Henry VIII of England" />

Reviewer: ```suggestion
<img src="/images/kinghenry.jpg" alt="King Henry VIII of England">
```
We implicitly close void elements - https://github.com/springernature/frontend-playbook/blob/master/markup/house-style.md#implicitly-close-void-elements

(I did all of these code examples and I put the closing `/` in there in all of them, it's all my fault 🤦‍♀ )

```

We don't do this:
```html

>  <img src="/images/kinghenry.jpg" alt="Image of King" />

Reviewer: ```suggestion
 <img src="/images/kinghenry.jpg" alt="Image of King">
```

```

### Meaningless images with meaningful alt attributes

If an image is decorative, the alt attribute should be present, but its value left blank. This conveys that the image isn’t necessary for understanding the page.

We do this:
```html

>  <img src="/images/spacer.gif" alt="" />

Reviewer: Oh god there's loads of them

```

###  Inappropriate negative tabindex

Interactive elements with a negative tabindex attribute (`tabindex=-1`) are not usable by keyboard users. 

### Visual order is not followed with source code
Just try to follow the visual order of elements while coding in HTML, don’t adjust elements to take positions with CSS, then this is confusing to keyboard users who expect to tab forwards and backwards in a linear sequence.

As a rule, try tabbing through your pages every so often just to make sure you haven't accidentally messed up the tab order. 

We do this:
```html
<button> I Should </button>
<button> Be focused</button>
<button> Last </button>
```

We don't do this:
```html
<button style=”float: right”> I Should </button>
<button> Be focused</button>
<button> Last </button>
```

## No advance warning when opening new window 

Opening new windows automatically when a link is activated can be disorienting for people who have difficulty perceiving visual content, and for some people with cognitive disabilities, if they are not warned in advance.

We do this:
```html
<a href="https://mysite.com" target="_blank" rel="noopener">
 My site

>  <span class="visuallyhidden">, opens in a new window</span>

Reviewer: Let's change this suggestion. A visually-hidden label doesn't do anything for sighted users, and sighted users can be confused by unexpected context changes too. 

The [a11y checklist recommends](https://github.com/springernature/frontend-playbook/blob/master/accessibility/accessibility-checklist.md#avoid-opening-links-in-new-windows) either a text label or an icon, so let's bring this remediation into line with that. 

I'm also going to open a PR on the checklist to clarify that the text label should be visible, as it's not clear enough there!
